### ISSUE TITLE
Rework new API for target runtimes

### ISSUE THREAD

**Issue by AndreyAkinshin** [dotnet/BenchmarkDotNet#1236]
In two recent pull requests (https://github.com/dotnet/BenchmarkDotNet/pull/1188 and https://github.com/dotnet/BenchmarkDotNet/pull/1230), new APIs for benchmark runtimes were introduced. In general, I like these changes: we are making steps in the right directions. However, I have a few major concerns which *should be resolved* before the next version of BenchmarkDotNet will be released (I already wrote my complaints [here](https://github.com/dotnet/BenchmarkDotNet/pull/1188#pullrequestreview-261540889) but I want to continue the discussion). Sorry for the late reaction, I had a pretty busy schedule lately.

## Smart obsoletes

We have a set of obsolete attributes like `ClrJob` or `CoreJob`. Currently, these attributes don't affect the config at all. IMHO, it's the worst way to make breaking changes because a BenchmarkDotNet version upgrade may introduce *unnoticeable* bugs. Such kind of obsoleting is even worse than removing these attributes (in this case, compile errors will force users to upgrade the API usages).

We should provide a discoverable way to learn how to upgrade the source code of the benchmarks. It's not enough to write about it in the release note and add a message in the Obsolete attribute (a lot of people just ignore it). I think that it's important because there are a lot of users who use these attributes (e.g., I found [600+ usages](https://github.com/search?l=C%23&q=ClrJob+-repo%3Adotnet%2FBenchmarkDotNet&type=Code) of `ClrJob` on GitHub). I see two possible ways to resolve this problem:

1. Introduce `InvalidConfigAttribute` and inherit all obsolete attributes from it. BenchmarkRunner should scan the target class for such attributes and print an error message instead of running benchmarks. The error message should contain information why did we do these changes and how to upgrade the benchmarks properly.
2. Implement backward-compatible behavior.

Personally, I like the second way.

## TargetFrameworkJobAttribute

I still think that the name of this attribute is poorly chosen. This name tells the reader that it's a job attribute which should have `TargetFrameworkMoniker` as a required parameter. But it's obvious because we can see this parameter as well. I'm trying to say that

```cs
[SimpleJob(TargetFrameworkMoniker.Mono)]
```

contains the same amount of useful information as

```cs
[TargetFrameworkJob(TargetFrameworkMoniker.Mono)]
```

But I prefer the first option because it contains less information noise (we don't have duplicated `TargetFramework` part) and it is easier to read it.
Also, this name is not consistent with other job attributes name. For example, we have the following signature for the `DryJob` constructor:

```cs
public DryJobAttribute(TargetFrameworkMoniker targetFrameworkMoniker, Jit jit, Platform platform)
```

And we still use the name `DryJob` instead of `DryTargetFrameworkJitPlatformJobAttribute`. And it's good: we shouldn't duplicate names of the ctor arguments in the attribute name. Instead of it, the attribute name should provide additional information which is not obvious from the usage (in this case, `Dry` tell us that the job is dry).

We can introduce a lot of job attributes which can be parametrized by `TargetFrameworkMoniker`, but there is no need to specify `TargetFramework` in the name of each attribute. I don't understand why `TargetFrameworkJobAttribute` is so special and why do we need it. It will be better to upgrade `SimpleJobAttribute`.

## API For Latest Runtime Version

As *a BenchmarkDotNet developer*, I understand why do we force users to specify target framework: it's easier for us to generate a csproj file for the generated benchmark. As *a BenchmarkDotNet users*, I'm unhappy about it because it makes our API worse:

1. This requirement looks pretty confusing. There is the only version of .NET Framework 4.x runtime installed on a Windows machine. A user wants to execute a benchmark on .NET Framework. Why should we force these users to specify the exact version of the .NET Framework despite the fact that the result will be the same?
2. It requires additional actions. A lot of users just don't know (or/and don't care) what the latest version of .NET Framework is installed on the machine. Modern IDEs allows creating new projects with the "default" target framework. But the current version of BenchmarkDotNet forces a user to check out the current version and specify it in the source code. The same for .NET Core: imagine that I want to execute a benchmark that doesn't have a runtime-specific logic (e.g., it just calls an external process). And I have a .NET Core versions on the machine. Why should I care about its version? Why BenchmarkDotNet forces me to specify this version?
3. Reduced portability. If a user uses `TargetFrameworkMoniker.Net48` as a target framework for a benchmark and share it on GitHub, it forces all of the future benchmark users to have .NET Framework 4.8 installed on the machine (or force them to spend time on the config patching).

I still think that we should have options for the latest versions (e.g., users should be able to ask for "Current .NET Framework" or "The latest installed stable .NET Core"). We already have logic that calculates the installed versions of .NET Framework and .NET Core. Why we can't just use it? I know that there are some corner cases, but in the case of a failure, we always can print a nice message and explain that it's recommended to specify the runtime version explicitly. I don't understand why we should spoil user experience for 100% of our users in order to avoid such situation for 1-5% users with corner case configurations.

## Future TargetFrameworkMoniker improvements

I want to continue discussing my proposal with renaming `TargetFrameworkMoniker`->`RuntimeMoniker`. @adamsitnik, in https://github.com/dotnet/BenchmarkDotNet/pull/1188#issuecomment-515871480, you wrote:

> I would prefer to keep the TargetFrameworkMoniker because this is what we use behind the scenes in auto-generated .csproj (except of Mono).

I don't think that we *shouldn't look at the implementation details* when we choose proper names for our API. Users should care about what do they want to achieve, not about how they want to achieve it. I think that from the user point of view, the following tasks look similar:

* I want to execute benchmarks on the installed .NET Framework
* I want to execute benchmarks on the latest installed stable version of .NET Core
* I want to execute benchmarks on .NET Core 2.1
* I want to execute benchmarks on the installed Mono
* I want to execute benchmarks in the current process

In all these sentences, a user wants to execute benchmarks on a specific runtime. I think that we should have a unified way to specify this runtime. Our API should help a user to express his/her desire. The internal implementation can be changed at any moment. E.g., the MSBuild team can rename `TargetFramework` to something else. We can invent a new way of benchmark running (e.g., we can find a way to create benchmark processes in memory instead of building explicitly generated csproj files; or we can send benchmarks to a remote server with proper .NET Runtime version and fetch the results). Any of these changes shouldn't affect our API because the user request will not be changed: our user wants to run benchmarks on a specific runtime.

I think that `RuntimeMoniker` is better than `TargetFrameworkMoniker` because:

1. It's shorter, and it contains a less number of words, which means that it's easy for reading.
2. It will not confuse users in the case of `TargetFrameworkMoniker.Mono` (there is no actual target framework called `Mono`).
3. It has fewer associations with .NET Framework (some users think that `TargetFramework` is something for .NET Framework only; I observe such a "false belief" all the time when discussing SDK-style projects with different developers).
4. It will allow introducing additional values (e.g., `NetCoreLatest` without additional points of confusing).

By the way, I still think that it's a good idea to introduce `RuntimeMoniker.InProcess`. Currently, we have `InProcessEmitToolchain.Instance` usage in our `IntroInProcess` which is an awful name: instead of thinking about the goal ("I want to run a benchmark in the current process"), we force users to think about internal BenchmarkDotNet entities. I think that all simple use cases (like running a benchmark in the current process) shouldn't use the `Toolchain` word at all: users shouldn't think about toolchains, it's an internal entity for BenchmarkDotNet developers. Instead of it, users should think about proper benchmark design; our goal is to provide a simple, laconic, and obvious API which will help them.

**Comment by adamsitnik** [comment]
> it's the worst way to make breaking changes because a BenchmarkDotNet version upgrade may introduce unnoticeable bugs

I've set error to true for all the obsolete attributes (the magic boolean attribute):

https://github.com/dotnet/BenchmarkDotNet/blob/a200af5878f808af6ade418b9a16f3650a11f572/src/BenchmarkDotNet.Annotations/Attributes/Obsolete/ClrJobAttribute.cs#L5-L6

It means that every BDN user is going to get a compilation **error** (not a warning) with a nice message that tells the user to use the new attribute.

**Comment by adamsitnik** [comment]
> I still think that the name of this attribute is poorly chosen. This name tells the reader that it's a job attribute which should have TargetFrameworkMoniker as a required parameter.

I agree, you are right. For some reason I've not thought about it and now when I see your proposal it's obvious to me.

Maybe we should go one step further and introduce `[JobAttribute]`? Or just extend the `[SimpleJobAttribute]`?

**Comment by adamsitnik** [comment]
> 
    I want to execute benchmarks on the installed .NET Framework
    I want to execute benchmarks on the latest installed stable version of .NET Core
    I want to execute benchmarks on .NET Core 2.1
    I want to execute benchmarks on the installed Mono
    I want to execute benchmarks in the current process

@AndreyAkinshin I am not sure if my changes were not too confusing. Specyfing the runtime is mandatory only if benchmark runtime != host runtime. 

I've explained that in https://github.com/dotnet/BenchmarkDotNet/blob/master/docs/_changelog/header/v0.11.6.md#multiple-frameworks-support

Again, all we need to do is just run the host process using the runtime we want, BDN recognizes current runtime and spawns the benchmarks using the same one.

**Comment by adamsitnik** [comment]
> TargetFrameworkMoniker->RuntimeMoniker

Naming is hard, honestly I am not sure which option to choose. Maybe I should just start a pool on Twitter to see what people like?

**Comment by adamsitnik** [comment]
> By the way, I still think that it's a good idea to introduce RuntimeMoniker.InProcess. 

I agree, we can extend the enum with two values:

* `Host` (default value)
* `InProcess`

> Instead of it, users should think about proper benchmark design; our goal is to provide a simple, laconic, and obvious API which will help them.

I totally agree. Honestly I think that we are learning API design while writing a library :D IMHO what @wojtpl2 have proposed in #1234 should help a lot with that. Also, I believe that we should put more focus on using console line arguments and fluent API. Attributes are fine as long as you have a single type with benchmarks.

**Comment by AndreyAkinshin** [comment]
## Obsoletes

> I've set error to true for all the obsolete attributes (the magic boolean attribute):
> It means that every BDN user is going to get a compilation error (not a warning) with a nice message that tells the user to use the new attribute.

Oh, didn't notice the `error` flag. Anyway, we have one more problem here: the current obsoleting approach prevents users from "smooth" upgrade: they will be forced to review all usages of such attributes and upgrade them with the new API. If they don't have enough time for it right now, the BenchmarkDotNet version upgrade will be postponed.
Check out [the statistics of BenchmarkDotNet downloads](https://www.nuget.org/stats/packages/BenchmarkDotNet?groupby=Version):
![image](https://user-images.githubusercontent.com/2259237/65829223-f5a89400-e2b3-11e9-9d07-5f3cc826b517.png)
We still have a lot of active BenchmarkDotNet v0.10.13 users. It's just hard to upgrade this version for some of them because we have some breaking changes here and there. I think that we should minimize the risk of "canceled upgrade" because of the compilation errors.
"Smooth" upgrade will be possible if we split each major breaking change into several phases:
* We obsolete old API with `error=false`, provide backward-compatible behavior, and introduce new API. In this case, users will be able to upgrade BenchmarkDotNet without additional changes in source code. They will have some time to review all obsolete usages after that.
* We obsolete old API with `error=true`. Here we will force users to upgrade all API usages, but it shouldn't be too surprising for them.
* We remove old API from our codebase.

## TargetFrameworkJobAttribute

> I agree, you are right. For some reason I've not thought about it and now when I see your proposal it's obvious to me.

Hooray! Let's change it!

> Maybe we should go one step further and introduce `[JobAttribute]`? Or just extend the `[SimpleJobAttribute]`?

I think it will be better to extend `[SimpleJobAttribute]`. IMHO, `[JobAttribute]` sounds too "generic". It feels like users should be able to define any kind of jobs with the help of `[JobAttribute]` (but it's not true). `[SimpleJobAttribute]` is more "honest" with users because its name says that it allows defining only "simple" jobs.

## API For Latest Runtime Version

> Specyfing the TFM is mandatory only for the users who want to test multiple runtimes (minority of the users). IMHO it's a fair price - users get a super powerfull feature for free and all they need to do is to specify TFM. When they want to test multiple runtimes they must know the monikers.

I agree that the feature is awesome. I'm not suggesting to remove this feature, I'm suggesting to extend it with special values that will be treated in a special way.

Let's say I have an SDK-style project with two target frameworks: `net48;netcoreapp30`. I already wrote these versions in the csproj file and now I want to benchmark my code with the help of BenchmarkDotNet. Why should I duplicate these versions in the BenchmarkDotNet config and write `RuntimeMoniker.Net48` and `RuntimeMoniker.NetCoreApp30`. The TFM version is important in csproj because it defines the API set that we can use in the current project. However, it doesn't really make sense to duplicate it in the BenchmarkDotNet config, it's an artificial limitation.

Let's say that we introduced `RuntimeMoniker`s like `NetLatest` and `NetCoreLatest`. Definitely, it will provide some benefits for users: better code portability (e.g., a code sample can be copy-pasted from a blog post or gist without additional actions), fewer duplications (between csproj and BDN config), easier setup (we shouldn't think about the current TFM version).
What kind of drawbacks will we get? Sometimes, users will have problems because it's impossible to correctly calculate the current TFM. Such problem can be easily detected (e.g., we can't build the generated benchmark or the real TFM from runtime doesn't equal to the calculated TFM), and we can warn users and say that they should provide the explicit TFM.
As a result, we will have an easy setup of most of users (who just want to benchmark two current TFMs) and one additional step for users who will have some problems with this feature (btw, nobody force them to use `NetLatest`, they can use existing explicit TFM versions like `Net48` without any limitations).

## Future TargetFrameworkMoniker improvements

> > TargetFrameworkMoniker->RuntimeMoniker
> Naming is hard, honestly I am not sure which option to choose. Maybe I should just start a pool on Twitter to see what people like?

The pool will not be representative if we don't provide the context and if we don't show pros and cons. I already listed 4 pros for `RuntimeMoniker` in the issue body. You can add pros for `TargetFrameworkMoniker` and start the pool (since Twitter has tweet length limitations, you can paste the description of our naming problem as an image).

> > By the way, I still think that it's a good idea to introduce RuntimeMoniker.InProcess.
> I agree, we can extend the enum with two values:
>    Host (default value)
>    InProcess

Looks very good to me!
(IMHO, these values add +1 for the `RuntimeMoniker` naming option because we don't have a lot of different values that are not real TargetFrameworkMoniker values; `Host`/`InProcess` makes sense only in the context of BenchmarkDotNet infrastructure).

> Honestly I think that we are learning API design while writing a library

Yep. =)

> Also, I believe that we should put more focus on using console line arguments and fluent API. Attributes are fine as long as you have a single type with benchmarks.

Agree. However, our primary goal right now is to finalize the current API changes and release `v0.12.0`. After that, we can start working on other improvements.

**Comment by AndreyAkinshin** [comment]
Resolved in #1277
